multidict/_multilib/pair_list.h

```diff
@@ -1132,23 +1132,45 @@ static inline int
 pair_list_update_from_pair_list(pair_list_t *list, PyObject* used, pair_list_t *other)
 {
     Py_ssize_t pos;
+    Py_hash_t hash;
+    PyObject *identity = NULL;
+    bool recalc_identity = list->calc_ci_indentity != other->calc_ci_indentity;
 
     for (pos = 0; pos < other->size; pos++) {
         pair_t *pair = other->pairs + pos;
+        if (recalc_identity) {
+            identity = pair_list_calc_identity(list, pair->key);
```

Comment:
The fix recalculates identity and its hash if the type of self does not equal the other's.